Clean up raise exception logic

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* **#55656 Clean up raise exception logic**
* #55237 Enable fast gradcheck for real inputs and outputs
* #56279 Disable flaky tests

### For release notes
What: 
 - All errors that are silenced by "raise_exception=False" are now GradcheckError (which inherits from RuntimeError).
 
Why:
 - Due to a refactor of gradcheck

Workaround:
 - If you catch for 'RuntimeError' with `except RuntimeError`, since GradcheckError inherits from RuntimeError, no changes are necessary. However if you explicitly check for the errors type via `type(error)`, you'll need to update your code to check for `GradcheckError` instead.

Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Differential Revision: [D27920809](https://our.internmc.facebook.com/intern/diff/D27920809)

soulitzer committed:
Clean up raise exception logic

[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"




[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -446,7 +438,7 @@ def backward_fn(grad_output):
     # Computes the analytical jacobian in slow mode for a single input_idx, output_idx pair
     # without performing checks for dtype, shape, and reentrancy
-    jacobians, _ = check_analytical_jacobian_attributes(inputs, outputs[output_idx], float('inf'), 1.0, False, False)
+    jacobians = check_analytical_jacobian_attributes(inputs, outputs[output_idx], float('inf'), 1.0, False)
```
You removed two arguments in the function signature: raise_exception, custom_vjp_fn=None
But only one here? 
I think we need to start passing these as kwargs!

soulitzer replied:
Its a bit confusing here for sure! I just wanted to remove custom_vjp_fn until we had some tests that actually need it

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1034,47 +1015,46 @@ def gradcheck(
+    if not raise_exception:
+        try:
+            return gradcheck_helper(**args)
+        except RuntimeError:
```
This simplification is really nice.
I think you can make it even better by defining a custom Error type here like `GradCheckError`. And only catch that error in this except. That way, you won't intercept any of the user errors

albanD reviewed (commented):
Just did a quick review, it looks mostly good.

Let me know once all the CI is good for a final review.

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Note: the except statement only catches RuntimeErrors so ValueErrors are still raised even if raise_exception=False, as expected.


[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1019,46 +1004,45 @@ def gradcheck(
-    tupled_inputs = _as_tuple(inputs)
+    # This is just a wrapper that handles the raise_exception logic
+    args = locals()
+    args.pop("raise_exception")
```
Does that modify the actual `locals` dict inplace? Or is it a shallow copy that you modify?

soulitzer replied:
We are actually modifying the locals dict inplace. At first that seemed okay, but 
from https://nedbatchelder.com/blog/201211/tricky_locals.html it looks like there are cases where the local symbol table is only copied into locals() after every statement. Good catch!

albanD reviewed (commented):
This is a great change! :)

I think we still want to mark this as BC-breaking though as we change the type of the error raised by gradcheck when it fails.
But that is really minimal.

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -499,7 +497,7 @@ def check_inputs(fail_test, tupled_inputs, check_sparse_nnz) -> bool:
             if content.layout is not torch._mkldnn:  # type: ignore
                 if not all(st > 0 or sz <= 1 for st, sz in zip(content.stride(), content.size())):
-                    raise RuntimeError(
+                    raise ValueError(
```
Oh right, this would be bc-breaking as well

soulitzer committed:
Update on "Clean up raise exception logic"


### For release notes
What: 
 - All errors that are silenced by "raise_exception=False" are now GradcheckError (which inherits from RuntimeError).
 
Why:
 - Due to a refactor of gradcheck

Workaround:
 - If you catch for 'RuntimeError' with `except RuntimeError`, since GradcheckError inherits from RuntimeError, no changes are necessary. However if you explicitly check for the errors type via `type(error)`, you'll need to update your code to check for `GradcheckError` instead.

Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.


[ghstack-poisoned]

albanD reviewed (approved):
LGTM!

soulitzer committed:
Update on "Clean up raise exception logic"


### For release notes
What: 
 - All errors that are silenced by "raise_exception=False" are now GradcheckError (which inherits from RuntimeError).
 
Why:
 - Due to a refactor of gradcheck

Workaround:
 - If you catch for 'RuntimeError' with `except RuntimeError`, since GradcheckError inherits from RuntimeError, no changes are necessary. However if you explicitly check for the errors type via `type(error)`, you'll need to update your code to check for `GradcheckError` instead.

Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.


[ghstack-poisoned]

soulitzer committed:
Update on "Clean up raise exception logic"


### For release notes
What: 
 - All errors that are silenced by "raise_exception=False" are now GradcheckError (which inherits from RuntimeError).
 
Why:
 - Due to a refactor of gradcheck

Workaround:
 - If you catch for 'RuntimeError' with `except RuntimeError`, since GradcheckError inherits from RuntimeError, no changes are necessary. However if you explicitly check for the errors type via `type(error)`, you'll need to update your code to check for `GradcheckError` instead.

Factors out all the logic handling involving `fail_test`, `raise_exception` into 1) a wrapper around gradcheck that uses try/except 2) gradcheck_helper that always raises exception.
This allows us to avoid having to write the `if not x: return False` logic that is scattered throughout gradcheck currently.

Differential Revision: [D27920809](https://our.internmc.facebook.com/intern/diff/D27920809)

[ghstack-poisoned]
